lib/ansible/plugins/inventory/gitlab_runners.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2018, Stefan Heitmueller <stefan.heitmueller@gmx.com>
+# Copyright (c) 2018 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: gitlab_runners
+    plugin_type: inventory
+    version_added: '2.8'
+    authors:
+      - Stefan Heitmüller (stefan.heitmueller@gmx.com)
+    short_description: Ansible dynamic inventory plugin for Gitlab runners.
+    requirements:
+        - python >= 2.7
+        - python-gitlab > 1.8.0
+    extends_documentation_fragment:
+        - constructed
+    description:
+        - Reads inventories from the Gitlab API.
+        - Uses a YAML configuration file gitlab_runners.[yml|yaml].
+    options:
+        plugin:
+            description: The name of this plugin, it should always be set to 'gitlab_runners' for this plugin to recognize it as it's own.
+            type: str
+            required: true
+            choices: gitlab_runners
+        server_url:
+            description: The URL of the Gitlab server, with protocol (i.e. http or https).
+            type: str
+            required: true
+            default: https://gitlab.com
+        api_token:
+            description: Gitlab token for logging in.
+            type: str
+            aliases:
+              - private_token
+              - access_token
+        filter:
+            description: filter runners from Gitlab API
+            type: str
+            choices: ['active', 'paused', 'online', 'specific', 'shared']
+'''
+
+EXAMPLES = '''
+# gitlab_runners.yml
+plugin: gitlab_runners
+host: https://gitlab.com
+
+# Example using constructed features to create groups and set ansible_host
+plugin: gitlab_runners
+host: https://gitlab.com
+strict: False
+keyed_groups:
+  # add e.g. amd64 hosts to an arch_amd64 group
+  - prefix: arch
+    key: 'architecture'
+  # add e.g. linux hosts to an os_linux group
+  - prefix: os
+    key: 'platform'
+  # create a group per runner tag
+  # e.g. a runner tagged w/ "production" ends up in group "label_production"
+  # hint: labels containing special characters will be converted to safe names
+  - key: 'tag_list'
+    prefix: tag
+'''
+
+from ansible.errors import AnsibleError
+from ansible.module_utils._text import to_native
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable
+
+try:
+    import gitlab
+    HAS_GITLAB = True
+except ImportError:
+    HAS_GITLAB = False
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable):
+    ''' Host inventory parser for ansible using Gitlab API as source. '''
+
+    NAME = 'gitlab_runners'
+
+    def _populate(self):
+        self.gl = gitlab.Gitlab(self.get_option('server_url'), private_token=self.get_option('api_token'))
```

Comment:
I don't know the gitlab API, but shouldn't you capture exceptions at this point?

Reply:
:see_no_evil: Sure and absolutely right.

Reply:
Ah, no initiating the object does not fail if invalid, it's the first request to the api like `gl.runners.all()` which fails.